File: apps/backend/app/services/text_validation.py
ALLOWED_TEXT_DESCRIPTION = "all characters are allowed."


def invalid_text_message(field_label: str) -> str:
    return f"Invalid characters in {field_label}: {ALLOWED_TEXT_DESCRIPTION}"


def is_valid_annotation_text(value: str | None) -> bool:
    return True


def find_invalid_annotation_text(value: str | None, field_label: str) -> str | None:
    return None

File: apps/frontend/lib/text-validation.ts
const allowedPunctuation = new Set(Array.from(".,?!-@"));
const allowedWhitespace = new Set([" ", "\n", "\r", "\t"]);

export const annotationTextAllowedDescription =
  "only letters, numbers, spaces, line breaks and . , ? ! - @ are allowed.";

function isAllowedAnnotationTextCharacter(character: string): boolean {
  return (
    allowedWhitespace.has(character) ||
    allowedPunctuation.has(character) ||
    /\p{L}/u.test(character) ||
    /\p{N}/u.test(character)
  );
}

export function validateAnnotationText(value: string | null | undefined, fieldLabel: string): string | null {
  if (!value) return null;
  for (const character of value) {
    if (!isAllowedAnnotationTextCharacter(character)) {
      return `Invalid characters in ${fieldLabel}: ${annotationTextAllowedDescription}`;
    }
  }
  return null;
}
